cumulus_library/studies/core/core_templates/encounter.sql.jinja

```diff
@@ -0,0 +1,117 @@
+two steps.
+
+TODO: profile speed vs a single step, consider extending date col methods
+to traverse nested elements if performance is impacted-#}
+CREATE TABLE core__encounter AS
+WITH temp_encounter_nullable AS (
+    SELECT DISTINCT
+        {{- utils.basic_cols(
+                'encounter',
+                'e',
+                [
+                    'id'
+                ], 
+            ) 
+        }},
+        {{- utils.nullable_cols(
+                'encounter',
+                'e',
+                [
+                    'status',
+                    'class', 
+                    ('subject', 'reference', 'subject_ref'),
+                    'period',
+                ], 
+                schema
+            ) 
+        }},
+        {{- utils.date_cols_from_str(
+                'encounter',
+                'e',
+                [
+                    ('period', 'start', 'period_start'),
+                ],
+                schema
+            ) 
+        }},
+        --{{ schema['encounter'] }} }}
+        {{- utils.truncate_date_cols(
+                'encounter',
+                'e',
+                [
+                    ('period','end', 'period_end_day', 'day'),
+                    ('period', 'start', 'period_start_day', 'day'),
+                    ('period', 'start', 'period_start_week', 'week'),
+                    ('period', 'start', 'period_start_month', 'month'),
+                    ('period', 'start', 'period_start_year', 'year'),
+                ],
+                schema
+            ) 
+        }},
+    FROM encounter AS e
+),
+
+temp_encounter AS (
+    SELECT DISTINCT
+        e.id,
+        e.status,
+        e.class,
+        e.subject_ref,
+        e.period_start,
+        e.period_start_day,
+        e.period_end_day,
+        e.period_start_week,
+        e.period_start_month,
+        e.period_start_year,
+        edt.code AS type_code,
+        edt.code_system AS type_code_system,
+        eds.code AS sevicetype_code,
+        eds.code_system AS sevicetype_code_system,
+        edp.code AS priority_code,
+        edp.code_system AS priority_code_system,
+        edr.code AS reasoncode_code,
+        edr.code_system AS reasoncode_code_system
+
+    FROM temp_encounter_nullable AS e
+    LEFT JOIN core__encounter_dn_priority AS edt ON e.id = edt.id
+    LEFT JOIN core__encounter_dn_servicetype AS eds ON e.id = eds.id
+    LEFT JOIN core__encounter_dn_priority AS edp ON e.id = edp.id
+    LEFT JOIN core__encounter_dn_reasoncode AS edr ON e.id = edr.id
+)
+
+SELECT DISTINCT
+    e.class AS enc_class,
+    ac.code AS enc_class_code,
+    ac.display AS enc_class_display,
+    e.status,
+    e.type_code,
+    e.type_code_system,
+    e.sevicetype_code,
+    e.sevicetype_code_system,
+    e.priority_code,
+    e.priority_code_system,
+    e.reasoncode_code,
+    e.reasoncode_code_system,
+    date_diff('year', date(p.birthdate), e.period_start_day) AS age_at_visit,
```

Comment:
This is probs fine since the ETL always gives you year-only birthDates. But if we ever want to handle cases where the day might actually fall anywhere in the year, this is the logic I used in `quality` (works in both athena and duckdb, unlike `date_diff`):

```
                year(targetDateTime)
                - year(birthDate)
                - (
                    CASE WHEN
                        extract(doy FROM birthDate) > extract(doy FROM targetDateTime)
                    THEN 1
                    ELSE 0
                    END
                )
```

Reply:
frankly i just copied this over verbatim from the raw sql without thinking about it.

but - can we talk about this? i assume there's a reason, which is why i didn't touch this, but that feels like an odd behavior.

Reply:
`date_diff` works like you want I believe, in Athena. (i.e. it takes day of year into account, is my memory)

But not duckdb - in duckdb, it is naive and just does year-minus-year.

So I used the logic above to get both to work the same. I'm guessing the author of that sql just didn't have duckdb in their brains because why would they. But you might hid oddities during unit testing if you ever feed this a non `-01-01` date.

Reply:
so the bit I actually wanted to talk about was truncating the birthdates at the ETL - just wanted to understand more of why that's happening.

Reply:
Oh that was because of reducing PHI vibes - we keep most datetimes in place, for clinical reasons. But I guess they (Andy & Jamie) felt like birthdates were more sensitive and we should keep truncating them.

For reference, that happens at the MS-deid-tool layer and we _could_ change the config to not do that.

Reply:
OK - I created #169 - we're using that now in a way that's slightly inaccurate and i want to, at some point, just discuss the implications once more.